File: FStudyForum.Infrastructure/Services/CategoryService.cs

using AutoMapper;
using FStudyForum.Core.Interfaces.IRepositories;
using FStudyForum.Core.Interfaces.IServices;
using FStudyForum.Core.Models.DTOs.Category;
using FStudyForum.Core.Models.Entities;
namespace FStudyForum.Infrastructure.Services;

public class CategoryService : ICategoryService
{

    private readonly ICategoryRepository _categoryRepository;
    private readonly IMapper _mapper;

    public CategoryService(
        ICategoryRepository categoryRepository,
        IMapper mapper)
    {
        _mapper = mapper;
        _categoryRepository = categoryRepository;
    }
    public async Task<List<CategoryDTO>> GetAllCategory()
    {
        var categories = await _categoryRepository.GetAll();
        var categoryDTOs = new List<CategoryDTO>();
        foreach (var categorie in categories)
        {
            categoryDTOs.Add(new CategoryDTO
            {
                Id = categorie.Id,
                Name = categorie.Name,
                Description = categorie.Description,
                Type = categorie.Type
            });
        }
        return categoryDTOs;
    }
    public async Task<CategoryDTO> CreateCategory(CreateCategoryDTO categoryDto)
    {
        if (await _categoryRepository.CateExists(categoryDto.Name))
        {
            var errorMessage = "Category Exist";
            throw new Exception(errorMessage);
        }
        var categories = new Category
        {
            Name = categoryDto.Name,
            Description = categoryDto.Description,
            Type = categoryDto.Type
        };
        var createdCategory = await _categoryRepository.Create(categories);
        await _categoryRepository.SaveChangeAsync();
        var createdCategoryDto = _mapper.Map<CategoryDTO>(createdCategory);
        return createdCategoryDto;
    }

    public async Task<CategoryDTO> GetCateByName(string name)
    {
       var category = await _categoryRepository.GetCateByName(name);
        if (category == null)
        {
            throw new Exception("Category not found");
        } 
        var categoryDto = new CategoryDTO
        {
            Id = category.Id,
            Name = category.Name,
            Description = category.Description,
            Type= category.Type 
        };

        return categoryDto;
    }

    public async Task<CategoryDTO> UpdateCategory(string name, UpdateCategoryDTO categoryDto)
{
    var existedCategory = await _categoryRepository.GetCateByName(name)
        ?? throw new Exception("Category not found");
    var categoryWithSameName = await _categoryRepository.GetCateByName(categoryDto.Name);
    if (categoryWithSameName != null && categoryWithSameName.Id != existedCategory.Id)
    {
        throw new Exception($"Category with name '{categoryDto.Name}' already exists.");
    }
    existedCategory.Name = categoryDto.Name;
    existedCategory.Description = categoryDto.Description;
    existedCategory.Type = categoryDto.Type;

    await _categoryRepository.Update(existedCategory);
    return _mapper.Map<CategoryDTO>(existedCategory);
}

    public async Task<bool> DeleteCategory(string name)
    {
         var existedCategory = await _categoryRepository.GetCateByName(name)
        ?? throw new Exception("Category not found");
        await _categoryRepository.Delete(existedCategory);
        return true;
    }
}

File: FStudyForum.Client/src/services/CategoryService.ts


import { ResponseWith } from "@/types/response";
import api from "./api";
import { Category, CreateCategoryDTO, UpdateCategoryDTO } from "@/types/category";

const getAllCategory = async (type?: string): Promise<Category[]> => {
  const response = await api.get<ResponseWith<Category[]>>(
    type ? `/category/all?type=${type}` : "/category/all"
  );
  return response.data.data;
};

const create = async (category: CreateCategoryDTO): Promise<Category> => {
  const response = await api.post<Category>("/category/create", category);
  return response.data;
};

const update = async (name: string, category: UpdateCategoryDTO): Promise<Category> => {
  const response = await api.put<Category>(`/category/update/${name}`, category);
  return response.data;
};

const deleteCategory = async (name: string): Promise<void> => {
  await api.delete(`/category/delete/${name}`);
};

const filterCategoriesByType = async (type: string): Promise<Category[]> => {
  const response = await api.get<ResponseWith<Category[]>>(`/category/filter`, { params: { type } });
  return response.data.data;
};

const CategoryService = {
  getAllCategory,
  create,
  update,
  deleteCategory,
  filterCategoriesByType
};

export default CategoryService;
